File: scripts/tools/schedule_pending_6_per_day.py
import json
import urllib.request
import os
import datetime

# Read .env.local
env = {}
with open('.env.local', 'r') as f:
    for line in f:
        if '=' in line and not line.startswith('#'):
            k, v = line.strip().split('=', 1)
            env[k] = v.strip('"')

supabase_url = env.get('NEXT_PUBLIC_SUPABASE_URL')
supabase_key = env.get('SUPABASE_SERVICE_ROLE_KEY') or env.get('NEXT_PUBLIC_SUPABASE_ANON_KEY')

url = f"{supabase_url}/rest/v1/posts?is_published=eq.false&select=id,title"
req = urllib.request.Request(url, headers={
    'apikey': supabase_key,
    'Authorization': f'Bearer {supabase_key}'
})

res = urllib.request.urlopen(req)
posts = json.loads(res.read())

print(f"Bắt đầu xếp lịch cho {len(posts)} bài...")

start_time = datetime.datetime.now()
start_time = start_time.replace(hour=8, minute=0, second=0, microsecond=0)

posts_in_day = 0
for post in posts:
    if posts_in_day >= 6:
        start_time += datetime.timedelta(days=1)
        start_time = start_time.replace(hour=8, minute=0, second=0, microsecond=0)
        posts_in_day = 0
    
    scheduled = start_time + datetime.timedelta(hours=posts_in_day*2)
    
    # Convert to UTC zero offset string compatible with postgres ISO
    iso_time = scheduled.strftime('%Y-%m-%dT%H:%M:%S.000Z')

    post_id = post['id']
    title = post['title'][:40]
    
    print(f"- Lên lịch: {title}... -> {scheduled.strftime('%d/%m/%Y %H:%M')}")
    
    # Update post - set scheduled_at for the auto-publish.mjs to pick up, and published_at for SEO date
    update_url = f"{supabase_url}/rest/v1/posts?id=eq.{post_id}"
    data = json.dumps({'scheduled_at': iso_time, 'published_at': iso_time, 'is_published': False}).encode('utf-8')
    update_req = urllib.request.Request(update_url, data=data, headers={
        'apikey': supabase_key,
        'Authorization': f'Bearer {supabase_key}',
        'Content-Type': 'application/json',
        'Prefer': 'return=minimal'
    }, method='PATCH')
    urllib.request.urlopen(update_req)
    
    posts_in_day += 1

print("✅ Đã hoàn tất cài đặt lịch 6 bài / ngày cho toàn bộ bài Nháp!")

File: scripts/tools/schedule_pending_6_per_day.mjs
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';

dotenv.config({ path: '.env.production' });
if (!process.env.NEXT_PUBLIC_SUPABASE_URL) {
  dotenv.config({ path: '.env.local' });
}

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

async function run() {
    console.log("Đang lấy 9 bài chưa đăng để phân bổ lịch...");
    
    const { data: posts, error } = await supabase
        .from('posts')
        .select('id, title, published_at, is_published')
        .eq('is_published', false)
        .order('id', { ascending: true });
        
    if (error) {
        console.error("Lỗi lấy bài:", error);
        return;
    }

    console.log(`Tìm thấy ${posts.length} bài. Đang xếp lịch 6 bài/ngày...`);

    let currentDate = new Date();
    // Bắt đầu xếp lịch từ ngày mai
    currentDate.setDate(currentDate.getDate() + 1); 
    currentDate.setHours(8, 0, 0, 0); // Bắt đầu từ 8h sáng

    let postsInDay = 0;

    for (let i = 0; i < posts.length; i++) {
        const post = posts[i];

        if (postsInDay >= 6) {
            // Đã đủ 6 bài cho ngày hiện tại, chuyển sang ngày tiếp theo
            currentDate.setDate(currentDate.getDate() + 1);
            currentDate.setHours(8, 0, 0, 0); // Về lại 8h sáng
            postsInDay = 0;
        }

        // Mỗi bài cách nhau 2 tiếng trong ngày (8h, 10h, 12h, 14h, 16h, 18h)
        const scheduledTime = new Date(currentDate);
        scheduledTime.setHours(8 + (postsInDay * 2));

        console.log(`- Xếp: "${post.title.substring(0, 40)}..." -> ${scheduledTime.toLocaleString('vi-VN')}`);

        const { error: updateError } = await supabase
            .from('posts')
            .update({ 
                published_at: scheduledTime.toISOString()
            })
            .eq('id', post.id);

        if (updateError) {
            console.error(`❌ Lỗi cập nhật ${post.id}:`, updateError);
        }

        postsInDay++;
    }
    
    console.log("✅ Đã hoàn tất lên lịch tự động!");
}

run();
